### PR TITLE
add KEEP_PACKAGE_SOURCES option to hunter_config()

### PR DESCRIPTION
**Pull request by headupinclouds**
PR for initial review/discussion:
* append `KEEP_PACKAGE_SOURCES` option in hunter_calculate_config_sha1.cmake if specified
* introduce _hunter_keep_package_sources in hunter_download.cmake as logical OR of global `HUNTER_KEEP_PACKAGE_SOURCES` and local `HUNTER_${h_name}_KEEP_PACKAGE_SOURCES`

### CODE REVIEW

File: cmake/modules/hunter_download.cmake [hunk 1]
```diff
@@ -429,7 +429,7 @@ function(hunter_download)
   file(
       APPEND
       "${HUNTER_DOWNLOAD_TOOLCHAIN}"
-      "set(HUNTER_KEEP_PACKAGE_SOURCES \"${HUNTER_KEEP_PACKAGE_SOURCES}\" CACHE INTERNAL \"\")\n"
+      "set(HUNTER_KEEP_PACKAGE_SOURCES \"${_hunter_keep_package_sources}\" CACHE INTERNAL \"\")\n"
```
Comment (ruslo) [inline cmake/modules/hunter_download.cmake]: Variable `_hunter_keep_package_source` is not defined at this point.

File: cmake/modules/hunter_calculate_config_sha1.cmake [hunk 2]
```diff
@@ -128,6 +128,11 @@ function(hunter_calculate_config_sha1 hunter_self hunter_base user_config)
         )
       endif()
 
+      string(COMPARE NOTEQUAL "{HUNTER_${x}_KEEP_PACKAGE_SOURCES}" "" have_keep_package_sources)
```
Comment (ruslo) [inline cmake/modules/hunter_calculate_config_sha1.cmake]: I don't see the point of changing `config-id`. Logically it's not affected.

File: cmake/modules/hunter_config.cmake [hunk 3]
```diff
@@ -87,4 +88,9 @@ macro(hunter_config)
   else()
     hunter_user_error("Expected VERSION option for 'hunter_config' command")
   endif()
+
+  if(_hunter_KEEP_PACKAGE_SOURCES)
+    set(HUNTER_${_hunter_current_project}_KEEP_PACKAGE_SOURCES ON)
```
Comment (headupinclouds) [inline cmake/modules/hunter_config.cmake]: LINK

File: cmake/modules/hunter_calculate_config_sha1.cmake (hunk 2 above)
Comment (headupinclouds) [reply cmake/modules/hunter_calculate_config_sha1.cmake]: I see.  The `set` call in `hunter_config. cmake` was being cl

When I had the following log at the bottom of `hunter_config.cmake`
```
  message("MOD: ${_hunter_current_project}  ${HUNTER_${_hunter_current_project}_KEEP_PACKAGE_SOURCES}")
```

It was clearing getting cleared on the third call.  

`grep MOD _logs/polly/default/log.txt  | grep GTest`

```
MOD: GTest
MOD: GTest  ON
MOD: GTest
```

So the package source wasn't being kept.  I noticed the third call to `hunter_config()` was coming from this generated file, and this mod fixed it:

```
MOD: GTest
MOD: GTest  ON
MOD: GTest  ON  # fixed
``` 

This all seems to be due to the misplaced variable in hunter_config.cmake [here](https://github.com/ruslo/hunter/pull/1216/files/13fdc60b17e81560b0c5e7cc0e24ebf53927ae36#r151868052).

**Comment by headupinclouds** [comment]
I've pushed a modified version up for discussion.  I'll prepare a minimal test to reproduce the issue I'm seeing.

File: cmake/modules/hunter_download.cmake (hunk 1 above)
Comment (ruslo) [reply cmake/modules/hunter_download.cmake]: All packages that depends on this package will be build with `HUNTER_KEEP_PACKAGE_SOURCES=ON` even if you want to keep sources only for one package. TL;DR I think this should be reverted.

File: cmake/modules/hunter_download.cmake [hunk 4]
```diff
@@ -382,6 +382,10 @@ function(hunter_download)
   file(REMOVE "${HUNTER_PACKAGE_HOME_DIR}/CMakeLists.txt")
   file(REMOVE "${HUNTER_DOWNLOAD_TOOLCHAIN}")
 
+  if(HUNTER_KEEP_PACKAGE_SOURCES OR HUNTER_${h_name}_KEEP_PACKAGE_SOURCES)
+    set(_hunter_keep_package_sources ON)
```
Comment (ruslo) [inline cmake/modules/hunter_download.cmake]: Let's add
```cmake
else()
  set(_hunter_keep_package_sources OFF)
endif()
```
just to be sure.

File: cmake/modules/hunter_download.cmake (hunk 1 above)
Comment (headupinclouds) [reply cmake/modules/hunter_download.cmake]: Reverted (will submit clean pr once it is working)

File: cmake/modules/hunter_download.cmake (hunk 4 above)
Comment (headupinclouds) [reply cmake/modules/hunter_download.cmake]: Changes applied [here](https://github.com/ruslo/hunter/pull/1216/commits/964c7fac0cada93077e51a22970d3f2fd4841afa)

The `HUNTER_${h_name}_KEEP_PACKAGE_SOURCES` is still getting cleared by the 3rd call to `hunter_config()` 

`grep "MOD: GTest" _logs/polly/libcxx/log.txt`

```
MOD: GTest
MOD: GTest  ON
MOD: GTest
```

which I believe comes from an evaluation of this call:

```
hunter_calculate_config_sha1.cmake:          "hunter_config(${x} "
```

Comment (ruslo) [reply cmake/modules/hunter_download.cmake]: > The HUNTER_${h_name}_KEEP_PACKAGE_SOURCES is still getting cleared by the 3rd call

It's not cleared - it wasn't set in scope we need (there is no `set(HUNTER_..._SOURCES "")` command or `unset(HUNTER_..._SOURCES)`).

TL;DR Use global property: https://github.com/ruslo/hunter/commit/c82aed035c3db6813d29306cc9e5b1f5b1a1fa06

```
MOD: GTest
MOD: GTest  ON
MOD: GTest
```

We are processing config 3 times:
* global default config: https://github.com/ruslo/hunter/blob/353628aee9e834e72c82ae9377c7fff019a11327/cmake/modules/hunter_calculate_config_sha1.cmake#L28-L30
* local config: https://github.com/ruslo/hunter/blob/353628aee9e834e72c82ae9377c7fff019a11327/cmake/modules/hunter_calculate_config_sha1.cmake#L36-L38
* unified config (merge of two configs which is used to calculate config-id and forwarding for all 3rd parties): https://github.com/ruslo/hunter/blob/353628aee9e834e72c82ae9377c7fff019a11327/cmake/modules/hunter_finalize.cmake#L136-L138

First two processed inside function so user's scope is not affected. We need something that will outlive local scope. It can be chain of `PARENT_SCOPE`, cache variables or properties. Global property fits nicely here.

Comment (headupinclouds) [reply cmake/modules/hunter_download.cmake]: Very helpful.  Thanks.

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.165
